main/com.spotify/spotifyd.js

```diff
@@ -0,0 +1,168 @@
+// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+// INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+// OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"use strict";
+
+const fs = require("fs");
+const path = require("path");
+const Tp = require("thingpedia");
+const child_process = require("child_process");
+const crypto = require("crypto");
+
+const SPOTIFYD_DIST_URL = "https://github.com/stanford-oval/spotifyd/releases/download/v0.3.1";
+
+function safeMkdir(dir) {
+    try {
+        fs.mkdirSync(dir);
+    } catch(e) {
+        if (e.code !== 'EEXIST')
+            throw e;
+    }
+}
+
+module.exports = class SpotifyDaemon {
+
+    constructor(options) {
+	if (!options) throw new Error("no options");
+        if (!options.cacheDir) throw new Error("no cacheDir");
+	if (!options.username) throw new Error("no username");
+	if (!options.device_name) throw new Error("no device name");
+	if (!options.token) throw new Error("no access token");
+	this.options = options;
+
+        this.cacheDir = options.cacheDir;
+	this.spotifydPath = path.join(this.cacheDir, "spotifyd", "spotifyd");
+	this._killed = false;
+	this._child = null;
+
+        this._init();
+    }
+
+    get_deviceId() {
+        return crypto.createHash('sha1').update(this.options.username).digest('hex'); // same protocol spotifyd uses
+    }
+
+    _checkIfInstalled() {
+        try {
+                return fs.existsSync(this.spotifydPath);
+        } catch (e) {
+            if (e.code == 'ENOENT') {
+                return false;
+            } else {
+                throw new Error(e);
+            }
+        }
+    }
+
+    async _download() {
+        safeMkdir(path.join(this.cacheDir, "spotifyd"));
+
+        let arch = ""
+        switch (process.arch) {
+            case 'x64':
+                break;
+            case 'armhf':
+                arch = "-armhf";
+                break;
+            default:
+                throw new Error("spotifyd unsupported architecture");
+        }
+
+        let url = SPOTIFYD_DIST_URL+"/spotifyd-linux"+arch+"-slim.tar.gz";
+        let destTempFile = path.join(this.cacheDir, "spotifyd.tar.gz");
+
+        try {
+            let r = child_process.spawnSync("wget", [ url, "-O", destTempFile ], { stdio: ['inherit', 'inherit', 'inherit'] });
+            if (r.status != 0) throw "wget";
+        } catch (e) {
+            throw new Error("failed to retrieve spotifyd binary package: " + e);
+        }
+
+        try {
+            let r = child_process.spawnSync("tar", [ "-xvf", destTempFile, "-C", path.join(this.cacheDir, "spotifyd") ]);
```

Comment:
Should we download the zip instead of tar? unzip is already a dependency of almond-server, while tar isn't.